services/ui-src/src/components/SynthesizedGrid/SynthesizedGrid.js

```diff
@@ -0,0 +1,104 @@
+import React, { useEffect, useState } from "react";
+import GridWithTotals from "../GridWithTotals/GridWithTotals";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+import { selectObjectInArrayByQuestionId } from "../../utilityFunctions/jsonPath";
+import { sortQuestionColumns } from "../../utilityFunctions/sortingFunctions";
+
+const SynthesizedGrid = props => {
+  const tempQuestionId = props.questionID;
+  const [synthGridData, setSynthGridData] = useState([]);
+  useEffect(() => {
+    const { answerData, gridData, allAnswers } = props;
+
+    let tempGridData = [];
+    const tabAnswers = allAnswers.filter(
+      element => element.rangeId === answerData.rangeId
+    );
+
+    answerData.rows.map((row, rowIndex) => {
+      // add header row
+      if (rowIndex === 0) {
+        tempGridData[rowIndex] = gridData[rowIndex];
+      } else {
+        let tempRowObject = {};
+        Object.entries(row).forEach(key => {
+          // get row header
+          if (key[0] === "col1") {
+            tempRowObject = Object.assign(tempRowObject, { [key[0]]: key[1] });
+          }
+          //calculate values for each column by row
+          else {
+            let tempCalculatedValue = calculateValue(key[1][0], tabAnswers);
+            tempRowObject = Object.assign(tempRowObject, {
+              [key[0]]: tempCalculatedValue
+            });
+          }
+        });
+        tempGridData.push(tempRowObject);
+      }
+      setSynthGridData(tempGridData);
+      return true;
+    });
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const calculateValue = (incomingCalculation, tabAnswers) => {
+    if (
+      incomingCalculation.actions &&
+      incomingCalculation.actions[0] === "formula"
+    ) {
+      let tempCalculation = [];
+      let returnValue = {};
+      Object.entries(incomingCalculation.targets).forEach(key => {
+        tempCalculation[key[0]] = getValue(key[1], tabAnswers);
+      });
+      // calculates the value based off of the formula <0> / <1>
+      returnValue = tempCalculation[0] / tempCalculation[1];
+      return returnValue;
+    }
+  };
+
+  const getValue = (target, tabAnswers) => {
```

Comment:
Does anyone have any suggestions on how this might be improved or be written so it is reusable?

Reply:
Abstracting it would make it better. You're looking for a specific row ([2]) in the array. If the data format changes, you have to rewrite the code, rather than the call. If you pass the headerrow as a parameter, for example, it would be less fragile. Tying it a specific data format like that makes it hard to reuse.